### ISSUE TITLE
Data packets sent to dropbear can be oversized

### ISSUE THREAD

**Issue by Deewiant** [ronf/asyncssh#459]
Sending large amounts of data to a process's stdin through dropbear causes dropbear to complain with "bad packet, oversized decompressed" and close the channel.

Checked with asyncssh v2.9.0 (current master) and dropbear v2020.81. I see no related changes in asyncssh `develop` and on the dropbear side it seems like it should reproduce on all versions from v2013.59 to current master — anything since https://github.com/mkj/dropbear/commit/49263b5314e59de654d7b01f513b1cd95b1b9ec9 .

Reproducing this should amount to just having dropbear running somewhere — in case it makes a difference, I've seen this with `dropbear -W 65536` which increases the default receive window size from 24576 to 65536 — and then e.g.:

```python
import asyncio
import asyncssh
import logging

logging.basicConfig()
asyncssh.set_log_level(logging.DEBUG)
asyncssh.set_debug_level(2)

async def run():
    client = await asyncssh.connect(...)
    data = "x" * 50000
    result = await client.run("cat", input=data)
    assert result.stdout == data

asyncio.get_event_loop().run_until_complete(run())
```

To be precise, this works up to a payload size of 32758 but fails for 32759 or greater. This is obviously no coincidence, from an excerpt of the log output:

```
DEBUG:asyncssh:[conn=0, chan=0] Set write buffer limits: low-water=16384, high-water=65536
INFO:asyncssh:[conn=0, chan=0] Requesting new SSH session
DEBUG:asyncssh:[conn=0, chan=0]   Initial recv window 2097152, packet size 32768
DEBUG:asyncssh:[conn=0, chan=0]   Initial send window 65536, packet size 32759
INFO:asyncssh:[conn=0, chan=0]   Command: cat
DEBUG:asyncssh:[conn=0, chan=0] Sending 32759 data bytes
DEBUG:asyncssh:[conn=0, chan=0] Sending EOF
DEBUG:asyncssh:[conn=0, chan=0] Reading from channel started
INFO:asyncssh:[conn=0] Connection failure: [Errno 104] Connection reset by peer
```

From what I can tell, dropbear negotiating a receive packet size of 32759 does not mean that it is willing to accept a `MSG_CHANNEL_DATA` with 32759 data bytes within it, but rather the whole `String(data)`, i.e. with the 4-byte length included, has to fit within those 32759 bytes, meaning there's room for only 32755 bytes in `data` itself. This feels somehow logical since it allows for 32768 bytes (dropbear's default `RECV_MAX_PAYLOAD_LEN`) to contain the 1-byte packet type, 4-byte channel, and in the case of `MSG_CHANNEL_EXTENDED_DATA` 4-byte data type, along with the rest of the payload. This snippet in dropbear's source suggests the same: https://github.com/mkj/dropbear/blob/a70b9e34c8119a82e05aaf4f7faa0ebf5a50dacc/common-channel.c#L62-L67

I don't know enough about the protocol to be able to say if this is actually an issue in asyncssh or in dropbear, but given the proliferation of old dropbear installations in embedded devices, it seems most practical to work around it in the client anyway.

**Comment by ronf** [comment]
RFC 4254 is not super clear on this, but here's what it says:

```
   The maximum amount of data allowed is determined by the maximum
   packet size for the channel, and the current window size, whichever
   is smaller.  The window size is decremented by the amount of data
   sent.  Both parties MAY ignore all extra data sent after the allowed
   window is empty.

   Implementations are expected to have some limit on the SSH transport
   layer packet size (any limit for received packets MUST be 32768 bytes
   or larger, as described in [SSH-TRANS].  The implementation of the
   SSH connection layer

   o  MUST NOT advertise a maximum packet size that would result in
      transport packets larger than its transport layer is willing to
      receive.

   o  MUST NOT generate data packets larger than its transport layer is
      willing to send, even if the remote end would be willing to accept
      very large packets.

   Additionally, some channels can transfer several types of data.  An
   example of this is stderr data from interactive sessions.  Such data
   can be passed with SSH_MSG_CHANNEL_EXTENDED_DATA messages, where a
   separate integer specifies the type of data.  The available types and
   their interpretation depend on the type of channel.

      byte      SSH_MSG_CHANNEL_EXTENDED_DATA
      uint32    recipient channel
      uint32    data_type_code
      string    data

   Data sent with these messages consumes the same window as ordinary
   data.
```

This suggests to me that the max packet size advertised in the open/open_confirmation messages is actual data bytes, not including any headers, including the length header that comes from encoding the data as a "string", as this length is meant to compare against the same length as the remaining window size, which definitely doesn't count any headers.

Looking at the definition of `RECV_MAX_CHANNEL_DATA_LEN` in the Dropbear code, it is reducing what it advertises by 9 bytes, which would be enough to allow space for the message type, channel, and the additional 4 bytes of length of the data string itself. It fact, there's a comment to that effect:

```c
/* allow space for:
 * 1 byte  byte      SSH_MSG_CHANNEL_DATA
 * 4 bytes uint32    recipient channel
 * 4 bytes string    data
 */
#define RECV_MAX_CHANNEL_DATA_LEN (RECV_MAX_PAYLOAD_LEN-(1+4+4))
```

It seems like that would be enough to avoid this problem, at lease on regular data packets. I'm not sure if they make an additional adjustment to account for the EXTENDED_DATA case, but in your sample code here you are just using stdin/stdout and not stderr, so I'm a bit surprised that you'd run into a problem, as long as AsyncSSH honors the smaller packet size of 32759 bytes (which it seems to from the debug output above).

How did `ret->size` end up greater than `RECV_MAX_PAYLOAD_LEN` in this case? Could it have something to do with compression, perhaps ending up with a larger compressed payload than the original payload being compressed? Do you see this problem with compression both enabled and disabled?

**Comment by Deewiant** [comment]
> How did `ret->size` end up greater than `RECV_MAX_PAYLOAD_LEN` in this case? Could it have something to do with compression, perhaps ending up with a larger compressed payload than the original payload being compressed? Do you see this problem with compression both enabled and disabled?

The error message saying "oversized decompressed" should have clued me in and I should've tested this more thoroughly to begin with. Indeed, the problem only occurs with compression enabled.

However, the payload is surely benefitting from compression here, since in the provided test case it's just a run of repeated `'x'` bytes. So it's not that the payload is expanding under compression, but something specific to how compression is handled.

I made the testing script more thorough, including having it start up dropbear with different settings. Here's a Dockerfile for isolated testing (based on Alpine just to reduce the download size a bit; at the moment the version of dropbear there is v2020.81):

```dockerfile
FROM python:3.10-alpine
RUN apk add dropbear gcc musl-dev libffi-dev
RUN pip install asyncssh
RUN mkdir /etc/dropbear && echo root:root | chpasswd
VOLUME /code
```

And the code:

```python
import asyncio
import asyncssh
import logging
import subprocess

logging.basicConfig()
asyncssh.set_log_level(logging.DEBUG)
asyncssh.set_debug_level(2)

server_up = asyncio.Event()
client_done = asyncio.Event()


async def run_server(recv_window):
    server = await asyncio.create_subprocess_shell(
        f"exec dropbear -F -R -E -W {recv_window} -p 127.0.0.1:22",
        stdin=subprocess.DEVNULL,
        stdout=subprocess.DEVNULL,
        stderr=None,
    )
    await asyncio.sleep(0.01)
    server_up.set()
    await client_done.wait()
    server.kill()
    await server.wait()


async def run_client(data_size, compression):
    await server_up.wait()
    kwargs = {}
    if not compression:
        kwargs["compression_algs"] = None
    client = await asyncssh.connect(
        "127.0.0.1", username="root", password="root", known_hosts=None, **kwargs
    )
    data = "x" * data_size
    result = await client.run("cat", input=data)
    client_done.set()
    return result.stdout == data


async def run_both(server, client):
    server_up.clear()
    client_done.clear()
    return await asyncio.gather(server, client)


async def main():
    sep = "  |  "
    header = sep.join(["compression", "dropbear recv window", "data size", "cat works"])
    results = []
    for compression in [True, False]:
        for recv_window in [32758, 32759]:
            for data_size in [recv_window - 1, recv_window, recv_window + 1, 100000]:
                print(
                    f"running: compression={compression}, recv_window={recv_window}, data_size={data_size}"
                )
                _, ok = await run_both(
                    run_server(recv_window), run_client(data_size, compression)
                )
                results.append(
                    sep.join(
                        [
                            "%11s" % compression,
                            "%20d" % recv_window,
                            "%9d" % data_size,
                            "%9s" % ok,
                        ]
                    )
                )
    print(header)
    for r in results:
        print(r)


asyncio.run(main())
```

Dropping this as `test.py` next to the Dockerfile and running `sudo docker run --rm -v $PWD:/code $(sudo docker build -q .) python /code/test.py`, the eventual results should be:

```
compression  |  dropbear recv window  |  data size  |  cat works
       True  |                 32758  |      32757  |       True
       True  |                 32758  |      32758  |       True
       True  |                 32758  |      32759  |       True
       True  |                 32758  |     100000  |       True
       True  |                 32759  |      32758  |       True
       True  |                 32759  |      32759  |      False
       True  |                 32759  |      32760  |      False
       True  |                 32759  |     100000  |      False
      False  |                 32758  |      32757  |       True
      False  |                 32758  |      32758  |       True
      False  |                 32758  |      32759  |       True
      False  |                 32758  |     100000  |       True
      False  |                 32759  |      32758  |       True
      False  |                 32759  |      32759  |       True
      False  |                 32759  |      32760  |       True
      False  |                 32759  |     100000  |       True
```

From playing around with these numbers a bit, it seems that as long as the dropbear receive window is 32759 or greater, sending 32759 or more bytes with compression enabled doesn't work. And the patch at #460 does appear to fix this.

**Comment by ronf** [comment]
Thanks very much for producing such a nice test case! I was able to get the Docker container to run here and reproduce your results above.

I still don't understand why #460 works as a fix, though, or if it's really addressing the root cause of the issue. I'll have to experiment with this some more.

**Comment by ronf** [comment]
I added some additional debugging on the AsyncSSH side to see what it is sending. It's definitely honoring the smaller dropbear packet size of 32759 bytes, and the data it ends up feeding to compress() is 32768 bytes as expected, after accounting for the packet type (1 byte), the channel (4 bytes), and the length of the payload (4 bytes). So, it seems like that would fit within the buffer size that dropbear should accept. As you said, the actual packets are much smaller than this. For instance, the 32768 byte data packet was 60 bytes after compression. Adding in padding, it came to 72 bytes including the 1-byte pad length, the compressed payload, and the padding.

The compression check in dropbear is specifically against RECV_MAX_PAYLOAD_LEN, which looks like it is 32768 here and that seems like it should be ok. However, I think I might have found the problem! The check is for greater than or equal, rather than greater than:

```python
		if (zstream->avail_out == 0) {
			int new_size = 0;
			if (ret->size >= RECV_MAX_PAYLOAD_LEN) {   # <----- This should be greater than!
				/* Already been increased as large as it can go,
				 * yet didn't finish up the decompression */
				dropbear_exit("bad packet, oversized decompressed");
			}
			new_size = MIN(RECV_MAX_PAYLOAD_LEN, ret->size + ZLIB_DECOMPRESS_INCR);
			ret = buf_resize(ret, new_size);
		}
```
So, when we send the full 32759 bytes plus overhead, we get to 32768 bytes exactly, and this trips.

It seems like even reducing the max payload size by 1 byte would probably be enough to fix this, but it's probably only necessary when communicating with dropbear, and probably only when compression is enabled based on your testing. I did a quick test with that change and got:

```
compression  |  dropbear recv window  |  data size  |  cat works
       True  |                 32758  |      32757  |       True
       True  |                 32758  |      32758  |       True
       True  |                 32758  |      32759  |       True
       True  |                 32758  |     100000  |       True
       True  |                 32759  |      32758  |       True
       True  |                 32759  |      32759  |       True
       True  |                 32759  |      32760  |       True
       True  |                 32759  |     100000  |       True
      False  |                 32758  |      32757  |       True
      False  |                 32758  |      32758  |       True
      False  |                 32758  |      32759  |       True
      False  |                 32758  |     100000  |       True
      False  |                 32759  |      32758  |       True
      False  |                 32759  |      32759  |       True
      False  |                 32759  |      32760  |       True
      False  |                 32759  |     100000  |       True
```

It's a bit of a hack, but since dropbear advertises itself in the client/server version string, it would be possible to have AsyncSSH automatically reduce its _send_pktsize by 1 when it detected this combination.

**Comment by ronf** [comment]
I'm thinking about a change that looks something like:

```patch
diff --git a/asyncssh/connection.py b/asyncssh/connection.py
index d753118..6543d23 100644
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -2411,6 +2411,11 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
         send_window = packet.get_uint32()
         send_pktsize = packet.get_uint32()

+        # Work around an off-by-one error in dropbear introduced in
+        # https://github.com/mkj/dropbear/commit/49263b5
+        if b'dropbear' in self._client_version and self._compressor:
+            send_pktsize -= 1
+
         try:
             chantype = chantype_bytes.decode('ascii')
         except UnicodeDecodeError:
@@ -2443,6 +2448,11 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
         send_window = packet.get_uint32()
         send_pktsize = packet.get_uint32()

+        # Work around an off-by-one error in dropbear introduced in
+        # https://github.com/mkj/dropbear/commit/49263b5
+        if b'dropbear' in self._server_version and self._compressor:
+            send_pktsize -= 1
+
         chan = self._channels.get(recv_chan)
         if chan:
             chan.process_open_confirmation(send_chan, send_window,
```

However, I can't seem to get a clean test run on the Docker container with this change. I'm getting an error of:

```
Traceback (most recent call last):
  File "/code/test.py", line 76, in <module>
    asyncio.run(main())
  File "/usr/local/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/code/test.py", line 58, in main
    _, ok = await run_both(
  File "/code/test.py", line 45, in run_both
    return await asyncio.gather(server, client)
  File "/code/test.py", line 24, in run_server
    server.kill()
  File "/usr/local/lib/python3.10/asyncio/subprocess.py", line 143, in kill
    self._transport.kill()
  File "/usr/local/lib/python3.10/asyncio/base_subprocess.py", line 153, in kill
    self._check_proc()
  File "/usr/local/lib/python3.10/asyncio/base_subprocess.py", line 142, in _check_proc
    raise ProcessLookupError()
ProcessLookupError
```

It's strange -- I wasn't seeing this failure when I unconditionally set the send_pktsize to be one byte smaller than what was requested. It's only showing up now that I'm doing it only when dropbear is detected and compression is enabled.

**Comment by Deewiant** [comment]
Great find! Your fix seems good to me. In theory a version/fork of dropbear that doesn't identify itself as such could exist out there, in which case the decrement would have to be unconditional (or user-controllable), but it's probably not worth worrying about.

(I remember now that as one of the first things I tried an off-by-one fix as well, but I think what happened is I did it like #460 but didn't realize to `break` from the loop, so it didn't work and I dismissed it, and eventually convinced myself of the off-by-four instead. Ah, well.)

I'm not sure what's going on with your Docker container, though. I applied your patch and it seems to work fine for me. Like so:

```sh
$ sudo docker run --rm -v $PWD:/code -it $(sudo docker build -q .) sh
$ apk add patch
$ patch /usr/local/lib/python3.10/site-packages/asyncssh/connection.py
<paste here>
$ python /code/test.py
```

This gets me a clean run of `True` for "cat works" like the one you showed earlier.

`ProcessLookupError` suggests that dropbear died on its own? I can't imagine why that would happen. (Relatedly, I noticed that I didn't close the client connection properly — the test code should really be doing `client.close()` and `await client.wait_closed()` in `run_client` — but doing that only causes the forked server-side subprocesses to die faster, and shouldn't affect the main process.)

For what it's worth, I can confirm that your fix also works for the real use case where I ran into this. So I don't think we really need to worry about problems in that hacky test script.

**Comment by ronf** [comment]
The above fix is now available as commit 37d0afc in the "develop" branch and will be include in the next release. Thanks again for helping me to get to the root of the issue!

**Comment by Deewiant** [comment]
Great, thanks for digging into it!

**Comment by ronf** [comment]
I just realized I left it out of the change log, but this fix has been included in AsyncSSH 2.10.1. Thanks again for reporting this and helping me to reproduce it. The updated changelog is now in the "develop" branch and will be included in the next release.
